File: hw/maisincho/u3/HW16MazeSolid/MazeJava/src/main/java/ec/edu/espe/mazegenerator/model/MazeBuilder.java

package ec.edu.espe.mazegenerator.model;

import java.util.Random;

/**
 *
 * @author Isaac Maisincho Crafters_Market DCCO ESPE
 */
public class MazeBuilder implements MazeGenerator {
        private final Random random = new Random();

    @Override
    public Maze generate(int width, int height) {
        Room[][] rooms = new Room[width][height];
        for (int x = 0; x < width; x++) {
            for (int y = 0; y < height; y++) {
                rooms[x][y] = new Room(x, y);
            }
        }

        Room entrance = rooms[0][0];
        Room exit = rooms[width - 1][height - 1];

        createPath(entrance, exit, rooms, width, height);
        return new Maze(rooms, entrance, exit);
    }

    private void createPath(Room current, Room exit, Room[][] rooms, int width, int height) {
        if (current.equals(exit)) {
            return;
        }

        int x = current.getX();
        int y = current.getY();

        if (x < width - 1) {
            Room right = rooms[x + 1][y];
            Door door = new Door(current, right);
            current.addDoor(door);
            right.addDoor(door);
            createPath(right, exit, rooms, width, height);
        } else if (y < height - 1) {
            Room down = rooms[x][y + 1];
            Door door = new Door(current, down);
            current.addDoor(door);
            down.addDoor(door);
            createPath(down, exit, rooms, width, height);
        }
    }
}

File: hw/maisincho/u3/HW17MazeSolid/MazePython/model/mazeBuilder.py
import random
from model.room import Room
from model.door import Door
from model.maze import Maze
from typing import List

class MazeBuilder:
    def generate(self, width: int, height: int) -> Maze:
        rooms: List[List[Room]] = [[Room(x, y) for y in range(height)] for x in range(width)]
        entrance = rooms[0][0]
        exit_room = rooms[width - 1][height - 1]
        self._create_path(entrance, exit_room, rooms, width, height)
        return Maze(rooms, entrance, exit_room)

    def _create_path(self, current: Room, exit_room: Room, rooms: List[List[Room]], width: int, height: int):
        if current == exit_room:
            return

        x, y = current.x, current.y

        if x < width - 1:
            right = rooms[x + 1][y]
            door = Door(current, right)
            current.add_door(door)
            right.add_door(door)
            self._create_path(right, exit_room, rooms, width, height)
        elif y < height - 1:
            down = rooms[x][y + 1]
            door = Door(current, down)
            current.add_door(door)
            down.add_door(door)
            self._create_path(down, exit_room, rooms, width, height)
